

# Project \#3726 Metagenomes of attached sulphur-cycling microbial communities in deep granitic groundwater of the Fennoscandian Shield#

## Overview ##

Three HiSeq-sequenced metagenomic samples from deep granitic groundwater have been assembled and QC-ed.

QC:

* Ran FastQC for general quality of the library
* Ran Trimmomatic to remove aapters and trim low quality reads

Assemblies:

* QCed reads have been assembled sample-wise using megahit (metabat in the proposed plan was a typo, metabat is a binning software) with and without digital normalization (total of 6 assemblies)
* sample-wise assembly with meta-spades (3 assemblies)
* co-assembly of all samples with megahit (not wit the normalized reads as the quality of the normalized assemblies was worse), and meta-spades with (2 assemblies)

This is a total of 11 assemblies.

Mapping:
"Usefullness/quality" of these assemblies has been be judged by mapping a subset of the reads to each assembly (filtered to 10kb , and 2.5kb), and percentage of mapped reads returned as indicator of quality.

Additionally standard assembly metrics have been computed.

## Processing ##

The data has been processed using a bash script calling a variety of tools including a custom python script. It is called 'this_script.sh' and can be found in the '999_deliverable'-folder. It will probably not run out of the box, but amending a number of file names and installing the right software might allow it to. It will however run for a reasonable amount of time (probably 2-3 weeks on a good desktop), and probably the metaspades coassembly will crash as it needed somewhere between 256 and 512Gb or RAM.

Needed software prerequisites:
* fastqc (https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
* trimmomatic (http://www.usadellab.org/cms/?page=trimmomatic)
* khmer (https://github.com/dib-lab/khmer)
* megahit (https://github.com/voutcn/megahit)
* spades (http://bioinf.spbau.ru/en/spades3.7)
* bbmap (https://sourceforge.net/projects/bbmap/)
* MASH (http://mash.readthedocs.io/en/latest/)

Additional a custom python helper script has been used, it can also be found in the '999_deliverable'-folder '999_deliverable'-folder ('helpers.py'). It uses a number of python packages to be used:
* Biopython
* pandas
* rpy2

## Data
The general structure of the project folder is as follows:
```
$base_path
├── 100_fastqc
│   ├── 110_prefiltering
│   └── 120_postfiltering
├── 200_cleaned_reads
├── 300_assemblies
│   ├── 301_all_assemblies
│   ├── 310_single_samples
│   │   ├── 311_megahit
│   │   │   ├── ONKKR15_S3
│   │   │   ├── ONKPVA06_S1
│   │   │   └── ONKPVA08_S2
│   │   ├── 312_normalised_megahit
│   │   │   ├── ONKKR15_S3
│   │   │   ├── ONKPVA06_S1
│   │   │   └── ONKPVA08_S2
│   │   └── 313_metaspades
│   │       ├── ONKKR15_S3
│   │       ├── ONKPVA06_S1
│   │       └── ONKPVA08_S2
│   └── 320_coassemblies
│       ├── 321_megahit
│       └── 322_metaspades
├── 400_maps
├── 500_assemblystats
│   └── 510_mash
└── 999_deliverable
```
All root folder correspond to main steps of the analysis and contain all the data generated by the corresponding tools. To be noted is the folder '301_all_assemblies' which contains all the assemblies at all the chosen length filters, and folder 400_maps which contains a folder for each assembly/sample combination.

## read QC and filtering

The results of the QC can be found in the folders in the '100_fastqc'-folder as '.html'-files. Eventhough the reports report some issues, I do not think anything particular is notable. It looks pretty OK, also most of the "bad" aspects are moderated by cleaning the reads with trimmomatic, or easily explainable by the metagenomic nature of the sample (fastQC does not assume this). The '200_cleaned_reads'-folder contains quality filtered and trimmed reads well as digitally normalized reads. Diverse suffixes describe the file type:

 The following table summarizes removed reads (all libraries are  compressed with gzip):
  * '_1P.gz' and '_2P.gz' : cleaned forward and reverse reads, when both of the pair passed
  * '_1U.gz' and '_2U.gz' : cleaned forward and reverse reads, when only one of the pair passed
  * '_clean.fastq.gz': interlazed cleaned reads, only pairs kept (e.g. same reads as '_1P' and '_2P')
  * '_normed.fastq.gz' : resulting reads from normalization of '_clean.fastq.gz'
  * '_normed_paired.fastq.gz' : same as before removing reads which lost their mate pair

Summary table of lost reads:

|library|reads in raw|reads in cleaned|removed reads|normalized reads|
|-------|------------|----------------|-------------|----------------|
|ONKKR15_S3|84 milion|80 milion|4 milion|13 milion|
|ONKPVA06_S1|58 milion|55 milion|3 milion|8 milion|
|ONKPVA08_S2|19 milion|18 milion|1 milion|2.6 milion|


## Assemblies

The '300_assemblies'-folder contains all the data generated by the assembly tools. The main results are in the '301_all_assemblies'-folder where I copied all the obtained assemblies in 'fasta'-format, as well as length filtered versions (in a self-explanatory naming scheme). I cleaned up some of the intermediate folders to reduce disk space but kept possibly informative filers.

Statistics on obtained assemblies are available in the table bellow, "prettier" versions (a shiny excel version, and a computer readable csv version) are available in the '500_assemblystats'-folder

|assembly                      |filter  | size      | nb_contig | min_contig_len | median_contig_len | mean_contig_len | max_contig_len | n50_len | n90_len | ONKKR15_S3 | ONKPVA06_S1 | ONKPVA08_S2 |
|----------------------------|-------|-----------|-----------|----------------|-------------------|-----------------|----------------|---------|---------|------------|-------------|-------------|
| megahit_ONKKR15_S3         | 0     | 413690946 | 292887    | 200            | 561               | 1412            | 926232         | 3313    | 472     | 94.827     | 53.5445     | 47.1865     |
| megahit_ONKKR15_S3         | 2500  | 227459025 | 24803     | 2501           | 4423              | 9170            | 926232         | 15173   | 3414    | 85.5205    | 43.4875     | 38.0495     |
| megahit_ONKKR15_S3         | 10000 | 135868079 | 4403      | 10001          | 16957             | 30858           | 926232         | 43450   | 12857   | 66.784     | 31.665      | 19.4805     |
| megahit_ONKPVA06_S1        | 0     | 443264135 | 320683    | 200            | 544               | 1382            | 772292         | 3231    | 457     | 54.443     | 94.6355     | 71.5335     |
| megahit_ONKPVA06_S1        | 2500  | 241376872 | 25740     | 2501           | 4360              | 9377            | 772292         | 16526   | 3425    | 44.992     | 84.6265     | 60.74       |
| megahit_ONKPVA06_S1        | 10000 | 147501089 | 4592      | 10003          | 17594             | 32121           | 772292         | 45707   | 13163   | 34.002     | 64.331      | 33.929      |
| megahit_ONKPVA08_S2        | 0     | 66443658  | 59596     | 200            | 505               | 1114            | 495440         | 1973    | 406     | 18.142     | 13.8275     | 94.407      |
| megahit_ONKPVA08_S2        | 2500  | 30700721  | 3142      | 2501           | 4543              | 9771            | 495440         | 18333   | 3509    | 9.46       | 6.839       | 87.3615     |
| megahit_ONKPVA08_S2        | 10000 | 19893556  | 705       | 10002          | 18945             | 28217           | 495440         | 34988   | 12919   | 6.302      | 4.5845      | 68.951      |
| megahit_coass              | 0     | 742438497 | 518627    | 200            | 556               | 1431            | 772274         | 3399    | 469     | 94.8865    | 94.648      | 92.3825     |
| megahit_coass              | 2500  | 414358242 | 46072     | 2501           | 4302              | 8993            | 772274         | 15207   | 3357    | 85.668     | 84.864      | 80.871      |
| megahit_coass              | 10000 | 244570725 | 7729      | 10001          | 17676             | 31643           | 772274         | 46063   | 13088   | 66.907     | 63.6375     | 53.211      |
| normed_megahit_ONKKR15_S3  | 0     | 419440006 | 299450    | 200            | 576               | 1400            | 326439         | 3085    | 475     | 94.9645    | 53.594      | 47.1885     |
| normed_megahit_ONKKR15_S3  | 2500  | 227435507 | 29789     | 2501           | 4461              | 7634            | 326439         | 10180   | 3265    | 85.9685    | 43.222      | 38.346      |
| normed_megahit_ONKKR15_S3  | 10000 | 115059226 | 5087      | 10001          | 15925             | 22618           | 326439         | 24506   | 11765   | 53.0645    | 26.259      | 15.895      |
| normed_megahit_ONKPVA06_S1 | 0     | 303628009 | 240048    | 200            | 533               | 1264            | 431703         | 2666    | 439     | 47.1935    | 92.522      | 66.5935     |
| normed_megahit_ONKPVA06_S1 | 2500  | 155264108 | 17809     | 2501           | 4549              | 8718            | 431703         | 13692   | 3385    | 34.6475    | 81.1295     | 48.579      |
| normed_megahit_ONKPVA06_S1 | 10000 | 92061074  | 3725      | 10007          | 16845             | 24714           | 431703         | 28423   | 12182   | 16.596     | 57.5605     | 15.478      |
| normed_megahit_ONKPVA08_S2 | 0     | 66905537  | 61934     | 200            | 509               | 1080            | 170908         | 1802    | 404     | 18.209     | 13.8405     | 93.634      |
| normed_megahit_ONKPVA08_S2 | 2500  | 29412150  | 3530      | 2501           | 4303              | 8332            | 170908         | 13545   | 3289    | 10.002     | 7.0345      | 86.213      |
| normed_megahit_ONKPVA08_S2 | 10000 | 16969488  | 686       | 10016          | 17419             | 24736           | 170908         | 28665   | 12573   | 7.286      | 4.699       | 59.0215     |
| spades_ONKKR15_S3          | 0     | 452509761 | 525007    | 55             | 322               | 861             | 808355         | 2487    | 277     | 95.773     | 55.0595     | 48.7235     |
| spades_ONKKR15_S3          | 2500  | 225973250 | 20449     | 2501           | 4755              | 11050           | 808355         | 22221   | 3785    | 90.6405    | 43.9425     | 40.74       |
| spades_ONKKR15_S3          | 10000 | 153192332 | 4537      | 10005          | 18249             | 33765           | 808355         | 51576   | 13635   | 79.864     | 33.889      | 29.0715     |
| spades_ONKPVA06_S1         | 0     | 550360732 | 751006    | 56             | 301               | 732             | 1332544        | 1541    | 264     | 56.5765    | 95.845      | 72.0645     |
| spades_ONKPVA06_S1         | 2500  | 240514511 | 21486     | 2501           | 4412              | 11194           | 1332544        | 27381   | 3672    | 45.952     | 88.067      | 65.0275     |
| spades_ONKPVA06_S1         | 10000 | 165072150 | 4342      | 10001          | 19793             | 38017           | 1332544        | 59677   | 14621   | 38.3945    | 79.1205     | 52.7025     |
| spades_ONKPVA08_S2         | 0     | 100056318 | 184580    | 55             | 281               | 542             | 725695         | 641     | 238     | 20.2645    | 15.6585     | 95.3995     |
| spades_ONKPVA08_S2         | 2500  | 30182518  | 2606      | 2501           | 4644              | 11581           | 725695         | 26821   | 3798    | 9.9875     | 7.112       | 89.9205     |
| spades_ONKPVA08_S2         | 10000 | 21302360  | 621       | 10035          | 20201             | 34303           | 725695         | 47417   | 14271   | 8.8735     | 6.042       | 82.062      |
| spades_coassembly          | 0     | 881695469 | 1049454   | 68             | 321               | 840             | 2450847        | 2294    | 275     | 96.001     | 96.062      | 94.404      |
| spades_coassembly          | 2500  | 431012329 | 39500     | 2501           | 4516              | 10911           | 2450847        | 24362   | 3693    | 91.316     | 89.51       | 88.4595     |
| spades_coassembly          | 10000 | 287930041 | 7721      | 10001          | 18983             | 37291           | 2450847        | 60728   | 14246   | 80.34      | 79.882      | 71.4505     |

## Discussion :

The main point I want to discuss is which assembly to continue with

I managed to run meta-SPAdes on all samples as well as a co-assembly, and the results are better than megahit in all cases (better mapping rates, longer contigs). Also the normalization of the data did not improve the assembly either. So I would go with the SPAdes assemblies. Except if you want to establish a more long lasting protocol and have limited computer resources, as the only draw-back I could see for the SPAdes assemblies is that they need significant computational ressources (mostly for the coassembly, I needed a 512Gb RAM machine, as opposed to all the megahit assemblies that I could run on a 64Gb RAM desktop machine).

As far as the SPAdes assemblies are concerned it will depend on your specific needs. The coassembly looks real good I have to admit, however the S2 library assemblies slightly better with the sample-wise assembly, mostly when filtering at 10kb (e.g. filtered at 2.5kb probably the gene content is the same but many 10kb+ contigs get split into smaller ones in the co-assembly).

Generally the sample-wise and the coassembly are of equivalent gene-content. However the co-assembly will allow a better direct comparison of the samples as there is only one reference for all three samples: it also has generally slightly longer contigs. However, there is a higher chance of certain types of miss-assemblies in the case of the coassembly, and also as already mentioned S2 seams to be doing a little bit worse.

For the level of filtering, I would recommend you to not take the unfiltered assemblies even though they show higher mapping rates as they can contain many (many, many, many) small contigs that while just create side effects in subsequent analyses. Choosing between 2.5kb and 10kb depends again on the downstream analyses. If only gene content is of importance 2.5kb if good enough (I would even go down to 1kb maybe, which I could do if you so desire. If you want accurate genome contexts however (e.g. linking genes to organisms) 10kb could be better and more accurate, people say. Personally I always used assemblies filtered at 1kb or 2.5kb, but it is basically up to you.


## A note on this document ##

This report has been written by Moritz Buck working (moritz.buck@nbis.se) for NBIS (National Bioinformatic Infrastructure Sweden) in the context of project \#3726.

For the whole document on environment variable will be used : `$base_path` which is the path to the folder in which all the analyses will be dumped.

## Resources used:

Human : 50hrs
Computer : 2800 hrs (I managed to run more then expected on my desktop machine)
